### PR TITLE
Added tag moderation to the member details view

### LINKED ISSUE #16522: Add the ability to add and remove moderated tags on the Admin Member Detailed View
📖 As an Admin, I want to be able to add and remove moderated tags for a user, so that I don't have to leave my current view and go to a separate page.

🐣 TASKS

- [ ]  Add the ability to add moderated tags directly from the Admin Member View 
- [ ] Add the ability to remove moderated tags directly from the Admin View

**Note this is an enhancement that will not hold back the deploy and release.**

### PR DESCRIPTION
**Pull request by nickytonline**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This introduces tag moderation at the user level in the member details view instead of having to navigate to a specific tag and then add a moderator based on their numeric user id.

cc: @michael-tharrington @ellativity @FounderCasey 

## Related Tickets & Documents

Closes #16522

## QA Instructions, Screenshots, Recordings

1. Run `bin/rails console`
2. In the rails console, enable the admin member view feature flag by running, `FeatureFlag.enable(:admin_member_view)`
3. exit the Rails console.
4. Login as the Forem admin account, email: admin@forem.local, password: password
5. Navigate to http://localhost:3000/admin/users and click on any username in the USER column

image

image

6. Start typing in the add tags input.

image

7. Tags should update for the user when saved.

https://user-images.githubusercontent.com/833231/154781522-c710b979-6ca7-4823-9cda-241f96c0cade.mp4

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [x] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by nickytonline** [comment]
The styling of selected tags is off because it currently deplends on https://github.com/forem/forem/blob/86f4000b6a85313daf3e2eb504b652f74fbac17d/app/assets/stylesheets/views/article-form.scss#L240-L304

Instead of having this tied directly to the article form, it would make sense to have this as a top level tag related CSS class for the buttons that are tags.

File: app/javascript/packs/admin/editUser.jsx [hunk 1]
```diff
@@ -1,4 +1,12 @@
+import { h, render } from 'preact';
+import { TagAutocompleteSelection } from '../../article-form/components/TagAutocompleteSelection';
```
Comment (nickytonline) [inline app/javascript/packs/admin/editUser.jsx:2]: These imports are currently specific to the article editor. Is there anything article editor specific about these @aitchiss or could they become part of the folder where the multi-autocomplete resides?

File: app/javascript/packs/admin/editUser.jsx [hunk 2]
```diff
@@ -81,15 +88,10 @@ const openModal = async (event) => {
   event.preventDefault();
 
   // Only load Preact if we haven't already.
-  if (!preact) {
-    [preact, { Modal: AdminModal }] = await Promise.all([
```
Comment (nickytonline) [inline app/javascript/packs/admin/editUser.jsx:85]: No longer dynamically importing Preact as we use need the `render` for the tag moderation section. The modal remains a dynamic import though.

File: app/javascript/packs/admin/editUser.jsx [hunk 3]
```diff
@@ -115,3 +117,28 @@ const openModal = async (event) => {
...
+const currentModTags = JSON.parse(
+  tagsRoot.closest('form')['current_mod_tags'].value,
+);
+
+render(
+  <MultiSelectAutocomplete
+    defaultValue={currentModTags}
+    fetchSuggestions={fetchTags}
+    labelText="Assign tags"
+    maxSelections={Math.MAX_VALUE}
```
Comment (nickytonline) [inline app/javascript/packs/admin/editUser.jsx:137]: I fell silly putting `Math.MAX_VALUE` at the moment. This could probably be something internal if no `maxSelections` prop is set @aitchiss ?

File: app/javascript/utilities/search/index.js [hunk 4]
```diff
@@ -144,3 +144,14 @@ export function fetchSearch(endpoint, dataHash) {
...
 }
+
+/**
+ * A helper method to search for tags.
+ *
+ * @param {string} query - A partial or complete tag name that is being searched for.
+ *
+ * @returns {Promise<object>} A collection of found tags, if any.
+ */
+export function fetchTags(name) {
```
Comment (nickytonline) [inline app/javascript/utilities/search/index.js:155]: I'm using this for this feature atm, but there are other places that can benefit from this helper function instead of using the generic `fetchSearch`. I may put this into a separate PR and update in the appropriate places.

File: app/javascript/packs/admin/editUser.jsx (hunk 1 above)
Comment (aitchiss) [reply app/javascript/packs/admin/editUser.jsx:2]: Yeah I guess we either move it to a shared component, or we move it to the multi-select autocomplete folder. The latter kinda suggests to me that we _always_ want to render tags like this (which we probably do?)

File: app/javascript/packs/admin/editUser.jsx (hunk 3 above)
Comment (aitchiss) [reply app/javascript/packs/admin/editUser.jsx:149]: `maxSelections` is an optional prop, and we don't limit selections if it's not passed in. I think you can just omit the prop here 🙂

File: app/javascript/packs/admin/editUser.jsx (hunk 1 above)
Comment (aitchiss) [reply app/javascript/packs/admin/editUser.jsx:2]: We can always move it again or rename it if we change our minds 😄

Comment (ludwiczakpawel) [reply app/javascript/packs/admin/editUser.jsx:2]: No thoughts here. We can probably move it all to live in one place.

File: app/javascript/packs/admin/editUser.jsx [hunk 5]
```diff
@@ -1,4 +1,44 @@
...
+    <MultiSelectAutocomplete
+      defaultValue={currentModTags}
+      fetchSuggestions={fetchTags}
+      labelText="Assign tags"
+      placeholder="Add a tag..."
+      showLabel={false}
+      SuggestionTemplate={TagAutocompleteOption}
+      SelectionTemplate={TagAutocompleteSelection}
+      onSelectionsChanged={syncSelections}
+      // onlySuggestions={true}
```
Comment (nickytonline) [inline app/javascript/packs/admin/editUser.jsx:36]: TODO: Sort this out. This was kind of working, but it breaks how the multi-select works. TLDR, I want to avoid adding new tags because a potential moderator can only moderate existing tags.TODO: Sort this out. This was kind of working, but it breaks how the multi-select works. TLDR, I want to avoid adding new tags because a potential moderator can only moderate existing tags.

Comment (aitchiss) [reply app/javascript/packs/admin/editUser.jsx]: Just an idea, but I think perhaps the `MultiSelectAutocomplete` shouldn't be responsible for this. 

So currently:
1. We pass in `fetchSuggestions` which returns only existing matching tags from the search call
2. In `MultiSelectAutocomplete` if the results are empty, we add the current search term to the suggestions

Instead we could:
- Not do point 2 above
- In `TagsField` which we use for the editor, we could make sure the `fetchSuggestions` prop we pass in adds the current search term as a suggestion if nothing was found in the search call

Comment (nickytonline) [reply app/javascript/packs/admin/editUser.jsx]: I got this working yesterday, my logic was wrong but let me try out your suggestion.

Comment (nickytonline) [reply app/javascript/packs/admin/editUser.jsx]: I've got it working as per your suggestion. Given the component only adds tags return from the `fetchSuggestions` prop now,  I'm going to alter some of the component tests as they assumed non-existent tags could be added.
